lib/ui/painting/image_decoder.cc

```diff
@@ -57,30 +62,91 @@ static SkISize GetResizedDimensions(SkISize current_size,
+    // image. There is nothing to do.
+    return image->makeRasterImage();
+  }
+
+  auto resized_dimensions =
       GetResizedDimensions(image->dimensions(), target_width, target_height);
 
   if (resized_dimensions.isEmpty()) {
     FML_LOG(ERROR) << "Could not resize to empty dimensions.";
     return nullptr;
   }
 
-  if (resized_dimensions == image->dimensions()) {
-    // The resized dimesions are the same as the intrinsic dimensions of the
-    // image. There is nothing to do.
-    return image;
+  // This is where the scaling happens.
+  if (resized_dimensions.width() > image->dimensions().width() ||
+      resized_dimensions.height() > image->dimensions().height()) {
+    FML_LOG(WARNING)
+        << "Image is being upsized. Using cache(Height|Width) correctly?";
+    // TOOD(jeltsch): consider exiting here, there's no good reason to support
+    // upsampling in a "caching"-optimiztion context..
   }
 
-  TRACE_EVENT0("flutter", __FUNCTION__);
-  flow.Step(__FUNCTION__);
+  const bool aspect_ratio_changed =
+      std::abs(AspectRatio(resized_dimensions) -
+               AspectRatio(image->dimensions())) > 0.01;
+  if (aspect_ratio_changed) {
+    // This is probably a bug. If a user passes dimensions that change the
+    // aspect ratio in a "caching" context that's probably not working as
+    // intended and rather a signal that the API is hard to use.
+    FML_LOG(WARNING) << "Aspect ratio is being changed.";
+  }
 
-  const auto scaled_image_info = image->imageInfo().makeWH(
-      resized_dimensions.width(), resized_dimensions.height());
+  if (codec != nullptr) {
+    auto decode_size = codec->getScaledDimensions(
+        std::max(static_cast<double>(resized_dimensions.width()) /
+                     image->dimensions().width(),
+                 static_cast<double>(resized_dimensions.height()) /
+                     image->dimensions().height()));
+
+    // If the codec supports efficient sub-sampling during decode use the
+    // supported size instead.
+    // Note that this won't respect the exact user-requested size but instead
+    // choose a supported approximation.
+    if (decode_size != image->dimensions()) {
+      const auto scaled_image_info =
+          image->imageInfo().makeDimensions(decode_size);
+      SkBitmap scaled_bitmap;
+      if (!scaled_bitmap.tryAllocPixels(scaled_image_info)) {
+        FML_LOG(ERROR) << "Could not allocate bitmap when attempting to scale.";
+        return nullptr;
+      }
+
+      if (codec->getPixels(scaled_bitmap.pixmap()) ==
+          SkCodec::Result::kSuccess) {
+        // Marking this as immutable makes the MakeFromBitmap call share
+        // the pixels instead of copying.
+        scaled_bitmap.setImmutable();
+
+        auto decoded_image = SkImage::MakeFromBitmap(scaled_bitmap);
+        if (!decoded_image) {
+          FML_LOG(ERROR)
+              << "Could not create a scaled image from a scaled bitmap.";
```

Comment:
Based on headerdocs, this shouldn't happen if `tryAllocPixels` above succeeded.  We should add an `FML_DCHECK(FALSE)` here.

Reply:
(Or perhaps just `FML_DCHECK(decoded_image)` above.

I'm also not sure of the implications of continuing at this point - why wouldn't we just return nullptr here?

Reply:
I added the FML_DCHECK as you requested and also pushed the "image = decoded_image" into the else branch to fall back to the original behavior and avoid nullptr derefs in the NDEBUG case.